### ISSUE TITLE
Publishing/Pulling GitHub Docker images is broken

### ISSUE THREAD

**Issue by atc0005** [atc0005/go-ci#1]
I followed the directions for [publishing packages](https://docs.github.com/en/packages/publishing-and-managing-packages/publishing-a-package) here by creating a Personal Access Token, then logging in with the local Docker daemon on my test VM. That worked fine and I was able to build and upload container images to docker.pkg.github.com.

I then updated a test workflow file in the atc0005/dnsc project to use the container image. This failed at the **Initialize containers** step:

```
##[error]Docker pull failed with exit code 1
/usr/bin/docker version --format '{{.Server.APIVersion}}'
'1.40'
Docker daemon API version: '1.40'
/usr/bin/docker version --format '{{.Client.APIVersion}}'
'1.40'
Docker client API version: '1.40'
/usr/bin/docker ps --all --quiet --no-trunc --filter "label=87c201"
/usr/bin/docker network prune --force --filter "label=87c201"
/usr/bin/docker network create --label 87c201 github_network_ce8740c1cecc40499ef90336213d1f90
a3d668d503c5b6f2fa5a6874eee818b1942e836ed75b2a0254f78ccb746bc2ee
/usr/bin/docker pull docker.pkg.github.com/atc0005/go-ci/go-ci-stable:latest
Error response from daemon: Get https://docker.pkg.github.com/v2/atc0005/go-ci/go-ci-stable/manifests/latest: no basic auth credentials
##[warning]Docker pull failed with exit code 1, back off 3.496 seconds before retry.
/usr/bin/docker pull docker.pkg.github.com/atc0005/go-ci/go-ci-stable:latest
Error response from daemon: Get https://docker.pkg.github.com/v2/atc0005/go-ci/go-ci-stable/manifests/latest: no basic auth credentials
##[warning]Docker pull failed with exit code 1, back off 2.853 seconds before retry.
/usr/bin/docker pull docker.pkg.github.com/atc0005/go-ci/go-ci-stable:latest
Error response from daemon: Get https://docker.pkg.github.com/v2/atc0005/go-ci/go-ci-stable/manifests/latest: no basic auth credentials
##[error]Docker pull failed with exit code 1
```

I'm still digging, but it appears that using containers in this way may not be supported?

- https://github.community/t/docker-pull-from-public-github-package-registry-fail-with-no-basic-auth-credentials-error/16358
- https://github.community/t/docker-from-within-actions/122602
- https://github.com/containerd/containerd/issues/3291

Still digging, but if this is the case, I'll need to abandon publishing efforts to the GitHub Packages repos and instead push to Docker Hub, quay.io or another Docker image repo.

**Comment by atc0005** [comment]
https://github.community/t/use-docker-images-from-github-package-registry/16135

**Comment by atc0005** [comment]
https://github.com/actions/starter-workflows/issues/66

**Comment by atc0005** [comment]
https://gist.github.com/ethomson/60e664ef09051cea66dada5d53c62e6d

**Comment by atc0005** [comment]
So far the workarounds I've found haven't really done what I'm looking for.

**Comment by atc0005** [comment]
https://github.com/Azure/docker-login

This logs into a registry, but only appears to take effect after any configured `container` setting for a job has already triggered. In short, it triggers too late to solve the original problem.

I may look more into this later if I adopt a workflow similar to this one:

https://github.com/rsyslog/rsyslog-pkg-ubuntu/blob/269ed115ed8ff9a1d8b12e751c8cfb410d9b5169/.github/workflows/ci.yml#L63

Short version: multiple `docker run` calls with commands executed within a `bash -c` block.

**Comment by atc0005** [comment]
From https://github.com/atc0005/go-ci/issues/4#issuecomment-1027790120:

> GitHub Packages Container registry went "GA" status on 2021-06-21.
> 
> Refs:
> 
> - https://github.blog/changelog/2021-06-21-github-packages-container-registry-is-generally-available/
> - https://github.blog/2021-06-21-github-packages-container-registry-generally-available/
> 
> Documentation:
> 
> - https://github.com/features/packages
> - https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry
> - https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
